modules/frontend/frontend.go

```diff
@@ -0,0 +1,227 @@
+		}
+		return next
+	})
+}
+
+type roundTripper struct {
+	next    http.RoundTripper
+	handler Handler
+}
+
+// NewRoundTripper merges a set of middlewares into an handler, then inject it into the `next` roundtripper
+func NewRoundTripper(next http.RoundTripper, middlewares ...Middleware) http.RoundTripper {
+	transport := roundTripper{
+		next:  next,
+	}
+	transport.handler = MergeMiddlewares(middlewares...).Wrap(&transport)
+	return transport
+}
+
+func (q roundTripper) RoundTrip(r *http.Request) (*http.Response, error) {
+	level.Info(util.Logger).Log("blockStart", r.URL.Query().Get("blockStart"), "blockEnd", r.URL.Query().Get("blockEnd"))
+	return q.handler.Do(r)
+}
+
+// Do implements Handler.
+func (q roundTripper) Do(r *http.Request) (*http.Response, error) {
+	level.Info(util.Logger).Log("msg", "roundTripper.Do called")
+	return q.next.RoundTrip(r)
+}
+
+func ShardingWare(shardNum int, logger log.Logger, registerer prometheus.Registerer) Middleware {
+	return MiddlewareFunc(func(next Handler) Handler {
+		return shardQuery{
+			next:     next,
+			shardNum: shardNum,
+			logger:   logger,
+			splitByCounter: promauto.With(registerer).NewCounterVec(prometheus.CounterOpts{
+				Namespace: "tempo",
+				Name:      "query_frontend_split_queries_total",
+				Help:      "Total number of underlying query requests after sharding",
+			}, []string{"user"}),
+		}
+	})
+}
+
+type shardQuery struct {
+	next     Handler
+	shardNum int
+	logger   log.Logger
+	// Metrics.
+	splitByCounter *prometheus.CounterVec
+}
+
+// Do implements Handler
+func (s shardQuery) Do(r *http.Request) (*http.Response, error) {
+	level.Info(s.logger).Log("msg", "shardQuery called")
+
+	userID, err := user.ExtractOrgID(r.Context())
+	if err != nil {
+		return nil, err
+	}
+
+	// create sharded queries
+	boundaryBytes := make([][]byte, s.shardNum+1)
+	for i := 0; i < s.shardNum+1; i ++ {
+		boundaryBytes[i] = make([]byte, 0)
+	}
+	const MaxUint = ^uint64(0)
+	const MaxInt = int64(MaxUint >> 1)
+	for i := 0; i < s.shardNum; i++ {
+		binary.PutVarint(boundaryBytes[i], MaxInt*(int64(i))/int64(s.shardNum))
+		binary.PutVarint(boundaryBytes[i], 0)
+	}
+	binary.PutVarint(boundaryBytes[s.shardNum], MaxInt)
+	binary.PutVarint(boundaryBytes[s.shardNum], MaxInt)
+
+	reqs := make([]*http.Request, s.shardNum)
+	for i := 0; i < s.shardNum; i++ {
+		reqs[i] = r
+		reqs[i].URL.Query().Add("blockStart", string(boundaryBytes[i]))
+		reqs[i].URL.Query().Add("blockEnd", string(boundaryBytes[i+1]))
+	}
+	s.splitByCounter.WithLabelValues(userID).Add(float64(s.shardNum))
+
+	rrs, err := DoRequests(r.Context(), s.next, reqs)
+	if err != nil {
+		return nil, err
+	}
+
+	// todo: add merging logic here if there are more than one results
+	for _, rr := range rrs {
+		if rr.Response.StatusCode == http.StatusOK {
+			return rr.Response, nil
+		}
+	}
+
+	return nil, fmt.Errorf("trace not found")
+}
+
+// RequestResponse contains a request response and the respective request that was used.
```

Comment:
All of this `RequestResponse` code seems copy/pasted out of Cortex.  These are all exported structs and methods.  Why don't we just use them directly?